### DISCUSSION TITLE
Question about Diarization inference

### DISCUSSION THREAD

**Discussion by naymaraq** [NVIDIA-NeMo/NeMo#5123] [Q&A] [answered]
Hi guys

I have couple of questions about Diarization inference script

1. https://github.com/NVIDIA/NeMo/blob/f964e1c17c3922d6f35b5e453aa712f2be153e6d/nemo/collections/asr/parts/utils/nmesc_clustering.py#L723 

Here, we are trying to marge clustering outputs together using ```torch.mode```.  But it seems wrong, because at each trial the output class indices are different. So majority voting will not produce any benefits. Can you explain how it works? 

https://github.com/NVIDIA/NeMo/blob/f964e1c17c3922d6f35b5e453aa712f2be153e6d/nemo/collections/asr/parts/utils/nmesc_clustering.py#L895
2. Why are you making graph fully connected after binarization? Are there any corner cases in which fully connected graph is necessary? 

Thanks

**Comment by okuchaiev** [comment, marked as answer]
@nithinraok any comments?

**Reply by tango4j** [reply]
1. At each seed, the function may generate a unique clustering output (oftentimes it is not unique.). We do have permutation issues but this part of code selects the majority labels including permutation. Try with `n_random_trials=30` and you will notice that the clustering result is more stable than `n_random_trials=1`. Note that the more random trials does not guarantee the better performance. 

2. If the affinity graph is not fully connected, there is a chance that number of speakers can be overcounted. Since overcounting of speakers in speaker diarization results in a significant error, the affinity graph is modified to have a fully connected, binarized affinity graph.

**Comment by naymaraq** [comment]
Thanks
